### PR TITLE
stm32h7: Add flash driver support

### PR DESCRIPTION
**Pull request by erwango**
Respawn of https://github.com/zephyrproject-rtos/zephyr/pull/30043 after it has been reverted due to issues detected in CI on flash driver handling on M4 core based targets.

Includes commits from #30043 + required commits to properly define flash nodes on M4 cores.
This fixes compilation issues when building flash driver on stm32h7 m4 target **BUT**:

Support of flash driver from M4 core requires more thinking, so as a consequence, its support is disabled:
- zephyr,flash-controller chosen declaration is removed for these cores
- A specific compilation error is generated when trying to build driver on this core

### CODE REVIEW

**Comment by erwango** [comment]
^^@nicovince Would you mind review and confirm this is fine for you ?

**Comment by nicovince** [comment]
Looks good to me.

**Review by ABOSTM — APPROVED** [review]
LGTM

**Comment by erwango** [comment]
@nicovince can you explicitly approve in a review, so this can be merged ?

**Review by nicovince — APPROVED** [review]
Approve changes

**Comment by ABOSTM** [comment]
@galak, is it possible to merge ?
I am waiting for this merge to push further commits.

File: drivers/flash/flash_stm32h7x.c [hunk 1]
```diff
@@ -0,0 +1,531 @@
+/*
+ * Copyright (c) 2020 Vossloh Cogifer
+ *
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+#include <kernel.h>
+#include <device.h>
+
+#define DT_DRV_COMPAT st_stm32h7_flash_controller
```
Comment (gmarull) [inline drivers/flash/flash_stm32h7x.c:10]: nit: compat is usually added on top before anything else

File: dts/arm/st/h7/stm32h743.dtsi [hunk 2]
```diff
@@ -7,6 +7,14 @@
 #include <st/h7/stm32h7.dtsi>
 
 / {
+	soc {
+		flash-controller@52002000 {
+			flash0: flash@8000000 {
+				write-block-size = <32>;
+				erase-block-size = <131072>;
```
[the remark this answers is not in the document]
Comment (erwango) [reply dts/arm/st/h7/stm32h743.dtsi]: Indeed, txs

**Review by gmarull — APPROVED** [review]

**Pull request merged**